extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,615 @@
+      rootBoundEnd - boundingRectStart : boundingRectEnd;
+    const percInView = lengthInView / (boundingRectEnd - boundingRectStart);
+    if (percInView < 0) {
+      return 0;
+    } else if (percInView > 1) {
+      return 1;
+    } else {
+      return percInView;
+    }
+  }
+
+  /**
+   * Handles serializing and sending messages to the safeframe.
+   * @param {!Object} payload
+   * @param {string} serviceName
+   * @private
+   */
+  sendMessage_(payload, serviceName) {
+    dev().assert(this.iframe_.contentWindow,
+        'Frame contentWindow unavailable.');
+    const message = dict();
+    message[MESSAGE_FIELDS.CHANNEL] = this.channel;
+    message[MESSAGE_FIELDS.PAYLOAD] = JSON.stringify(
+        /** @type {!JsonObject} */(payload));
+    message[MESSAGE_FIELDS.SERVICE] = serviceName;
+    message[MESSAGE_FIELDS.SENTINEL] = this.sentinel_;
+    message[MESSAGE_FIELDS.ENDPOINT_IDENTITY] = this.endpointIdentity_;
+    this.iframe_.contentWindow./*OK*/postMessage(
+        JSON.stringify(message), SAFEFRAME_ORIGIN);
+  }
+
+  /**
+   * Routes messages to their appropriate handler.
+   * @param {!JsonObject} payload
+   * @param {string} service
+   */
+  processMessage(payload, service) {
+    switch (service) {
+      case SERVICE.CREATIVE_GEOMETRY_UPDATE:
+        this.handleFluidMessage_(payload);
+        break;
+      case SERVICE.EXPAND_REQUEST:
+        this.handleExpandRequest_(payload);
+        break;
+      case SERVICE.REGISTER_DONE:
+        this.handleRegisterDone_(payload);
+        break;
+      case SERVICE.COLLAPSE_REQUEST:
+        this.handleCollapseRequest_();
+        break;
+      default:
+        break;
+    }
+  }
+
+  /**
+   * @param {!JsonObject} payload
+   * @private
+   */
+  handleRegisterDone_(payload) {
+    this.initialHeight_ = payload['initialHeight'];
+    this.initialWidth_ = payload['initialWidth'];
+  }
+
+  /**
+   * @param {!JsonObject} payload
+   * @private
+   */
+  handleExpandRequest_(payload) {
+    let expandHeight = payload['expand_b'] + payload['expand_t'];
+    let expandWidth = payload['expand_r'] + payload['expand_l'];
+    // We assume that a fair amount of usage of this API will try to pass the final
+    // size of the iframe that is desired, instead of the correct usage which is
+    // specify how much to expand by. We try to protect against this, by detecting
+    // if the requested expansion size matches the size of the amp-ad element, and
+    // if so we assume that the desired effect was to resize to fill the element.
+    if (expandWidth != this.initialWidth_ ||
+        expandHeight != this.initialHeight_) {
+      expandWidth += Number(this.iframe_.width);
+      expandHeight += Number(this.iframe_.height);
+    }
+    this.handleSizeChange(expandHeight,
+        expandWidth,
+        SERVICE.EXPAND_RESPONSE);
+  }
+
+  /**
+   * @private
+   */
+  handleCollapseRequest_() {
+    this.handleSizeChange(this.initialSize_.height,
```

Comment:
Should we check that there was a proceeding expand request?  I believe the logic doesn't exist as it would cause the iframe to shrink

Reply:
The container code will not allow an expand to be sent if not collapsed, and vice versa. A second collapse wouldn't even do anything.